File: app/src/main/java/com/morozione/roboblog/database/ImageDao.java
//package com.morozione.roboblog.database;
//
//import android.graphics.Bitmap;
//import android.graphics.BitmapFactory;
//import android.net.Uri;
//import android.os.Handler;
//import android.util.Log;
//
//import com.google.firebase.storage.FirebaseStorage;
//import com.google.firebase.storage.StorageMetadata;
//import com.google.firebase.storage.StorageReference;
//import com.google.firebase.storage.UploadTask;
//import com.maxml.offirent.BuildConfig;
//import com.maxml.offirent.utils.OffirentConstants;
//
//import java.io.ByteArrayOutputStream;
//import java.io.File;
//import java.io.IOException;
//import java.util.ArrayList;
//import java.util.List;
//
//public class ImageDao  {
//    private StorageReference storageRef;
//
//    public ImageDao(Handler handler) {
//        super(handler);
//
//        if (storageRef == null) {
//            storageRef = FirebaseStorage.getInstance().getReference();
//        }
//    }
//
//    public void load(String name) {
//        File localFile = null;
//        try {
//            localFile = File.createTempFile("Images", "jpg");
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//        final File finalLocalFile = localFile;
//        if (localFile == null)
//            return;
//
//        storageRef.child(name).getFile(localFile)
//                .addOnSuccessListener(taskSnapshot -> {
//                    Bitmap image = BitmapFactory.decodeFile(finalLocalFile.getAbsolutePath());
//                    List<Bitmap> list = new ArrayList<>();
//                    list.add(image);
//                    success(OffirentConstants.HANDLER_IMAGE_LIST, list);
//                }).addOnFailureListener(e -> {
//        });
//    }
//
//    public void save(String uri) {
//        StorageMetadata metadata = new StorageMetadata
//                .Builder()
//                .setContentType("image/jpg")
//                .build();
//
//        Uri file = Uri.fromFile(new File(uri));
//        final String s = "image-" + file.getLastPathSegment().hashCode() + "-" + file.getLastPathSegment();
//        UploadTask uploadTask = storageRef.child(s).putFile(file, metadata);
//
//        uploadTask.addOnProgressListener(taskSnapshot -> {
//            double progress = (100.0 * taskSnapshot.getBytesTransferred()) / taskSnapshot.getTotalByteCount();
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "onProgress " + (int) progress);
//        }).addOnSuccessListener(taskSnapshot -> {
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "onSuccess Https " + taskSnapshot.getDownloadUrl());
//            success(OffirentConstants.IMAGEDAO_RESULT_PATH_OK, s);
//        }).addOnFailureListener(e -> {
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "onFailure " + e.toString());
//        });
//    }
//
//    public void saveBitmap(Bitmap bitmap, String nameImage) {
//        final String path = "image-" + nameImage;
//        ByteArrayOutputStream baos = new ByteArrayOutputStream();
//        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, baos);
//        byte[] data = baos.toByteArray();
//
//        UploadTask uploadTask = storageRef.child(path).putBytes(data);
//        uploadTask.addOnProgressListener(taskSnapshot -> {
//            double progress = (100.0 * taskSnapshot.getBytesTransferred()) / taskSnapshot.getTotalByteCount();
//            success(OffirentConstants.IMAGEDAO_UPLOAD_PROGRESS, progress);
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "onProgress " + (int) progress);
//        }).addOnFailureListener(e -> {
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "onError loading: " + e.getMessage());
//        }).addOnSuccessListener(taskSnapshot -> {
//            if (BuildConfig.DEBUG)
//                Log.i(OffirentConstants.LOG_TAG, "image loated");
//            if (taskSnapshot.getMetadata() == null ||
//                    taskSnapshot.getMetadata().getDownloadUrl() == null) {
//                return;
//            }
//            String temp = taskSnapshot.getMetadata().getDownloadUrl().toString();
//            success(OffirentConstants.IMAGEDAO_RESULT_PATH_OK, temp);
//        });
//    }
//}

File: app/src/main/java/com/morozione/roboblog/database/ImageDao.kt
package com.morozione.roboblog.database

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import com.google.firebase.Firebase
import com.google.firebase.storage.StorageMetadata
import com.google.firebase.storage.storage
import io.reactivex.rxjava3.core.Single
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException

class ImageDao : BaseDao() {
    private var storageRef = Firebase.storage.reference

    fun load(name: String): Single<Bitmap> {
        return Single.create { emitter ->
            try {
                val localFile = File.createTempFile("Images", "jpg")
                storageRef.child(name)
                    .getFile(localFile)
                    .addOnSuccessListener {
                        val bitmap = BitmapFactory.decodeFile(localFile.absolutePath)
                        if (bitmap != null) {
                            emitter.onSuccess(bitmap)
                        } else {
                            emitter.onError(Exception("Failed to decode bitmap"))
                        }
                        // Clean up temp file
                        localFile.delete()
                    }
                    .addOnFailureListener { emitter.onError(it) }
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }
    }

    fun save(uri: String): Single<String> {
        return Single.create { emitter ->
            try {
                val metadata = StorageMetadata.Builder()
                    .setContentType("image/jpg")
                    .build()

                val file = Uri.fromFile(File(uri))
                val fileName = "image-${file.lastPathSegment?.hashCode()}-${file.lastPathSegment}"
                
                storageRef.child(fileName)
                    .putFile(file, metadata)
                    .addOnSuccessListener { taskSnapshot ->
                        taskSnapshot.metadata?.reference?.downloadUrl
                            ?.addOnSuccessListener { downloadUrl ->
                                emitter.onSuccess(downloadUrl.toString())
                            }
                            ?.addOnFailureListener { emitter.onError(it) }
                            ?: emitter.onError(Exception("Failed to get download URL"))
                    }
                    .addOnFailureListener { emitter.onError(it) }
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }
    }

    /**
     * Save bitmap to Firebase Storage
     * Note: EXIF rotation should be handled by ImageUtil before calling this method
     */
    fun saveBitmap(bitmap: Bitmap, nameImage: String): Single<String> {
        return Single.create { emitter ->
            try {
                val path = "image-$nameImage"
                val baos = ByteArrayOutputStream()
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, baos) // Reduced quality for smaller size
                val data = baos.toByteArray()
                baos.close()

                storageRef.child(path)
                    .putBytes(data)
                    .addOnSuccessListener { taskSnapshot ->
                        taskSnapshot.metadata?.reference?.downloadUrl
                            ?.addOnSuccessListener { downloadUrl ->
                                emitter.onSuccess(downloadUrl.toString())
                            }
                            ?.addOnFailureListener { emitter.onError(it) }
                            ?: emitter.onError(Exception("Failed to get download URL"))
                    }
                    .addOnFailureListener { emitter.onError(it) }
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }
    }
    
    /**
     * Save bitmap with proper orientation handling - alternative method 
     * This method can be used if ImageUtil rotation doesn't work properly
     */
    fun saveBitmapWithRotation(bitmap: Bitmap, nameImage: String, originalPath: String? = null): Single<String> {
        return Single.create { emitter ->
            try {
                val path = "image-$nameImage"
                
                // Apply rotation if needed based on original path
                val correctedBitmap = originalPath?.let { path ->
                    try {
                        val exif = ExifInterface(path)
                        val orientation = exif.getAttributeInt(
                            ExifInterface.TAG_ORIENTATION,
                            ExifInterface.ORIENTATION_NORMAL
                        )
                        
                        when (orientation) {
                            ExifInterface.ORIENTATION_ROTATE_90 -> rotateBitmap(bitmap, 90f)
                            ExifInterface.ORIENTATION_ROTATE_180 -> rotateBitmap(bitmap, 180f)
                            ExifInterface.ORIENTATION_ROTATE_270 -> rotateBitmap(bitmap, 270f)
                            else -> bitmap
                        }
                    } catch (e: IOException) {
                        bitmap // Return original bitmap if EXIF reading fails
                    }
                } ?: bitmap
                
                val baos = ByteArrayOutputStream()
                correctedBitmap.compress(Bitmap.CompressFormat.JPEG, 90, baos)
                val data = baos.toByteArray()
                baos.close()

                storageRef.child(path)
                    .putBytes(data)
                    .addOnSuccessListener { taskSnapshot ->
                        taskSnapshot.metadata?.reference?.downloadUrl
                            ?.addOnSuccessListener { downloadUrl ->
                                emitter.onSuccess(downloadUrl.toString())
                            }
                            ?.addOnFailureListener { emitter.onError(it) }
                            ?: emitter.onError(Exception("Failed to get download URL"))
                    }
                    .addOnFailureListener { emitter.onError(it) }
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }
    }
    
    /**
     * Rotate bitmap by specified degrees
     */
    private fun rotateBitmap(bitmap: Bitmap, degrees: Float): Bitmap {
        val matrix = Matrix().apply {
            postRotate(degrees)
        }
        
        return try {
            Bitmap.createBitmap(
                bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true
            )
        } catch (e: OutOfMemoryError) {
            bitmap // Return original bitmap if rotation fails due to memory issues
        }
    }
}